Initial doctest impl

chmp opened the pull request:
As discussed in #223, I think testing the documentation requires a custom plugin. At the moment it's very rough around the edges in particular when it comes to error reporting. 

When I run the current state, I already get the following errors:

```
FAILED tests/test_docs.py::test_docexample[docexample2] - NameError: name 'wasm2wat' is not defined
FAILED tests/test_docs.py::test_docexample[docexample8] -   File "foo", line 6
FAILED tests/test_docs.py::test_docexample[docexample15] - NameError: name 'Module' is not defined
FAILED tests/test_docs.py::test_docexample[docexample17] - AssertionError
FAILED tests/test_docs.py::test_docexample[docexample20] - AssertionError
FAILED tests/test_docs.py::test_docexample[docexample25] - NameError: name 'Store' is not defined
FAILED tests/test_docs.py::test_docexample[docexample28] - NameError: name 'Store' is not defined
FAILED tests/test_docs.py::test_docexample[docexample40] - ModuleNotFoundError: No module named 'wasmer_compiler_llvm'
FAILED tests/test_docs.py::test_docexample[docexample44] - ImportError: cannot import name 'FunctionTYpe' from 'wasmer' (/mnt/c/Users/cmp/Documents/Code/SideProjects/wasmer-python/.env/lib/python3.8/site-packages/wasmer.cpython-38-x86_64-linux-gnu.so)
FAILED tests/test_docs.py::test_docexample[docexample49] - ModuleNotFoundError: No module named 'wasmer_compiler_llvm'
FAILED tests/test_docs.py::test_docexample[docexample50] - pyo3_runtime.PanicException: Nor `clang-10` or `clang` has been found, at least one of them is required for the `NativeEngine`: CannotFindBinaryPath
FAILED tests/test_docs.py::test_docexample[docexample55] - AttributeError: 'builtins.StateBuilder' object has no attribute 'generate_import_object'
FAILED tests/test_docs.py::test_docexample[docexample61] - RuntimeError: preopened directory not found: `foo`
FAILED tests/test_docs.py::test_docexample[docexample62] - RuntimeError: preopened directory not found: `foo`
```

One major limitation is that there are no stacktraces:

```
____________________________________________________________________________________________________________________________ test_docexample[docexample61] _____________________________________________________________________________________________________________________________

docexample = <conftest.DocTest object at 0x7ff7d7bcf310>

    def test_docexample(docexample):
>       docexample.exec()

tests/test_docs.py:2:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
conftest.py:60: in exec
    exec(co)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   RuntimeError: preopened directory not found: `foo`

foo:4: RuntimeError
_______________________
```

[My past experiments](https://cprohm.de/article/better-test-output-with-ast-rewriting-and-a-patched-standard-library.html/) however lead me to believe that's also not achievable without monkey patching the standard library, as IPython is doing. TBH. I don't think the result woudl justify the effort. Maye adding a custom exception with a more explanatory repr would already be enough.

Hywan commented:
Haha, you're on fire! Thank you very much, I highly appreciate!

> One major limitation is that there are no stacktraces

I don't care to be honest. We just need to ensure that tests are up-to-date and valid, so that users can copy-paste them safely.

chmp commented:
> Haha, you're on fire! Thank you very much, I highly appreciate!

I have been using wasmer in some projects and enjoyed it a lot, so I though why not contribute back.

> I don't care to be honest. 

TBH, my immediate concern is also to figure out what I need to change in the docs to the get the checks to pass :D

chmp commented on tests/test_docs.py:
```diff
@@ -0,0 +1,20 @@
+
+import wasmer
+
+known_issues = {
```
Anything that took me longer than 5 secs to understand, I skipped. When running pytest with `-rxXs` should [show the details](https://docs.pytest.org/en/latest/how-to/skipping.html).

Hywan replied:
Good to me. I can try to fix them. The PR is already an awesome improvement!

chmp commented:
- I added the source of the test to the error output. It's now quite verbose, but otherwise it was really hard to find the corresponding lines in the Rust source
- I implemented a rudimentary, hard-coded xfail mechansim for the doctests
- I fixed some documentation errors, but skipped everything that I did not understood right away. The tests are xfailed so, they are still executed.

Overall, I think that should do it.

Hywan commented on conftest.py:
```diff
@@ -0,0 +1,79 @@
+        self.start_line = start_line
+
+    def exec(self):
+        # make sure the line numbers between exception and  source agree
```
```suggestion
        # make sure the line numbers between exception and source agree

```

Hywan commented on conftest.py:
```diff
@@ -0,0 +1,79 @@
+
+
+def pytest_generate_tests(metafunc):
+    if 'docexample' in metafunc.fixturenames:
```
Can you rename `docexample` to `doctest` please? That's usually the common vocabulary.

Hywan commented:
I've merged this PR manually. I've updated the following things:

* I've changed CRLF to LF in new files you added,
* I've renamed `docexample` to `doctest`, and I've improved the parameterized test IDs, https://github.com/wasmerio/wasmer-python/commit/d2010a7fda7aff86597a5627c62c491bade8e02d (now the IDs are the `str` representation of the `obj`)
* I've fixed all the known issues:
  * https://github.com/wasmerio/wasmer-python/commit/1e047cf8bfa44ab311f508f819110d32bbf02157
  * https://github.com/wasmerio/wasmer-python/commit/7a1315b230144d7993a084c63c550e3efa4b9dac
  * https://github.com/wasmerio/wasmer-python/commit/2c7dbad54d5b128083bbdba96a9e8e990af9ba66
  * https://github.com/wasmerio/wasmer-python/commit/a338b7b5f66b73f637ffb1ab346f64ff5178c355
  * https://github.com/wasmerio/wasmer-python/commit/94fe9a9573e25bbc17d1496cf6d4cfcb37d9c3ea
  * https://github.com/wasmerio/wasmer-python/commit/19bffcb37e742665d334feffda03528da85a1242
  * https://github.com/wasmerio/wasmer-python/commit/ea906331f29f08f5270300c2bc455a25c27bef73

Hywan commented:
Thank you very much, it's really cool to get this feature!

Hywan commented:
We would need to update our `pdoc3` fork to support `py,ignore` tag for docblock.

```
```py,ignore
```

chmp commented:
Thanks for the feedback. I'm really happy, that the changes help. 

One last thing, I thought of: depending on your preference, you could try to move the conftest.py file into the tests folder to declutter the root. I think if you do not rely on proper plugins it should work.
